sound/soc/sof/pcm.c

```diff
@@ -211,16 +211,22 @@ static int sof_pcm_hw_free(struct snd_soc_component *component,
 	dev_dbg(component->dev, "pcm: free stream %d dir %d\n",
 		spcm->pcm.pcm_id, substream->stream);
 
-	/* free PCM in the DSP */
-	if (pcm_ops && pcm_ops->hw_free && spcm->prepared[substream->stream]) {
-		ret = pcm_ops->hw_free(component, substream);
-		if (ret < 0)
-			err = ret;
+	if (spcm->prepared[substream->stream]) {
+		/* stop DMA first if needed */
+		if (pcm_ops && pcm_ops->delayed_platform_trigger)
```

Comment:
remind me if it's possible NOT to have a pcm_ops?

Reply:
only for the DSP-less mode. I did not not use sdev->dsplessmod because those patches are not upstream and would block upstreaming this patch

Reply:
right exactly

Reply:
correct. There will be no case where we have duplicate stop triggers. We stop only once either during suspend or during hw_free